File: executor.go
// Runs `rg --json` and folds the stream into per-file records.
// Memory is O(files), never O(matches): counters stay exact while only a
// handful of representative lines per file are retained.

package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"sort"
	"strings"
	"unicode/utf8"
)

type Line struct {
	N     uint64
	Text  string
	Score float32
	IsDef bool
}

type FileHit struct {
	Path    string
	Matches uint64
	// Bit i set when token i of the query was seen anywhere in this file.
	// Tracked over every match line, so coverage is exact even though only a
	// few lines are retained.
	Covered uint32
	// Best representative lines, highest score first.
	Best []Line
	// Best definition-shaped line seen in this file, if any. Tracked over the
	// whole stream so a definition can never be lost to the retention cap.
	Def   *Line
	Score float32
}

type Outcome struct {
	Files      []*FileHit
	TotalMatch uint64
	TotalFiles uint64
	// Byte size of the equivalent `rg -n --no-heading` output. This is the
	// faithful baseline the never_worse invariant compares against.
	FaithfulBytes uint64
	RgExit        int
}

const (
	keepPerFile  = 3
	maxLineChars = 600
)

func runeLen(s string) int { return utf8.RuneCountInString(s) }

// takeRunes returns the first n runes of s and whether s was longer.
func takeRunes(s string, n int) (string, bool) {
	if utf8.RuneCountInString(s) <= n {
		return s, false
	}
	i, count := 0, 0
	for i < len(s) && count < n {
		_, size := utf8.DecodeRuneInString(s[i:])
		i += size
		count++
	}
	return s[:i], true
}

func stripPrefix(p string) string { return strings.TrimPrefix(p, "./") }

func digits(n uint64) int {
	if n == 0 {
		return 1
	}
	d := 0
	for n > 0 {
		d++
		n /= 10
	}
	return d
}

type rgEvent struct {
	Type string `json:"type"`
	Data *struct {
		Path *struct {
			Text *string `json:"text"`
		} `json:"path"`
		Lines *struct {
			Text *string `json:"text"`
		} `json:"lines"`
		LineNumber *uint64 `json:"line_number"`
	} `json:"data"`
}

func runRg(rgArgs []string, tokens []string, detectDefs bool) (*Outcome, error) {
	args := append([]string{"--json"}, rgArgs...)
	cmd := exec.Command("rg", args...)
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, err
	}
	var stderr strings.Builder
	cmd.Stderr = &stderr
	if err := cmd.Start(); err != nil {
		return nil, err
	}

	out := &Outcome{}
	var cur *FileHit
	// ReadString rather than Scanner: a single minified line can exceed any
	// fixed token cap, and dropping it would corrupt the counters.
	r := bufio.NewReaderSize(stdout, 1<<20)
	for {
		raw, err := r.ReadString('\n')
		if len(raw) > 0 {
			var ev rgEvent
			if json.Unmarshal([]byte(strings.TrimRight(raw, "\n")), &ev) == nil && ev.Data != nil {
				switch ev.Type {
				case "begin":
					path := ""
					if ev.Data.Path != nil && ev.Data.Path.Text != nil {
						path = *ev.Data.Path.Text
					}
					cur = &FileHit{Path: stripPrefix(path)}
				case "match":
					if cur == nil {
						break
					}
					// Non-UTF8 lines arrive as `bytes` (base64); we count them
					// but never try to display them.
					if ev.Data.Lines == nil || ev.Data.Lines.Text == nil {
						cur.Matches++
						out.TotalMatch++
						break
					}
					text := *ev.Data.Lines.Text
					var n uint64
					if ev.Data.LineNumber != nil {
						n = *ev.Data.LineNumber
					}
					cur.Matches++
					out.TotalMatch++
					out.FaithfulBytes += uint64(len(cur.Path) + 1 + digits(n) + 1 + len(text))

					trimmed := strings.TrimSpace(strings.TrimRight(text, "\n\r"))
					if trimmed == "" {
						break
					}
					display, cut := takeRunes(trimmed, maxLineChars)
					if cut {
						display += "…"
					}
					lower := strings.ToLower(display)
					for i, t := range tokens {
						if i >= 32 {
							break
						}
						if strings.Contains(lower, t) {
							cur.Covered |= 1 << uint(i)
						}
					}
					score, isDef := lineScore(display, tokens)
					isDef = isDef && detectDefs
					l := Line{N: n, Text: display, Score: score, IsDef: isDef}

					if isDef && (cur.Def == nil || l.Score > cur.Def.Score) {
						d := l
						cur.Def = &d
					}
					if len(cur.Best) < keepPerFile {
						cur.Best = append(cur.Best, l)
						sortBest(cur.Best)
					} else if l.Score > cur.Best[len(cur.Best)-1].Score {
						cur.Best[len(cur.Best)-1] = l
						sortBest(cur.Best)
					}
				case "end":
					if cur != nil {
						if cur.Matches > 0 {
							out.TotalFiles++
							out.Files = append(out.Files, cur)
						}
						cur = nil
					}
				}
			}
		}
		if err != nil {
			if err != io.EOF {
				return nil, err
			}
			break
		}
	}

	if err := cmd.Wait(); err != nil {
		if ee, ok := err.(*exec.ExitError); ok {
			out.RgExit = ee.ExitCode()
		} else {
			return nil, err
		}
	}
	if out.RgExit == 2 && strings.TrimSpace(stderr.String()) != "" {
		os.Stderr.WriteString(stderr.String())
	}
	return out, nil
}

func sortBest(b []Line) {
	sort.SliceStable(b, func(i, j int) bool { return b[i].Score > b[j].Score })
}

// passthrough streams rg verbatim to stdout, so output is byte-identical to
// plain rg. Used for passthrough and for the never_worse fallback.
//
// One deviation, and it is deliberate: the stream stops at `cap` bytes. An
// unanchored regex like `[a-z]{2,}` produces 18 MB on a mid-size repo, and
// handing that to an agent is worse than any ranking mistake. The cut is stated
// on stderr with the exact command to get the rest.
func passthrough(rgArgs []string, capBytes int) (int, error) {
	cmd := exec.Command("rg", rgArgs...)
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return 2, err
	}
	if err := cmd.Start(); err != nil {
		return 2, err
	}
	r := bufio.NewReaderSize(stdout, 1<<20)
	w := bufio.NewWriter(os.Stdout)

	written := 0
	truncated := false
	for {
		buf, err := r.ReadString('\n')
		if len(buf) > 0 {
			if written+len(buf) > capBytes {
				truncated = true
				break
			}
			if _, werr := w.WriteString(buf); werr != nil {
				break
			}
			written += len(buf)
		}
		if err != nil {
			break
		}
	}
	w.Flush()

	if truncated {
		_ = cmd.Process.Kill()
		fmt.Fprintf(os.Stderr,
			"\nscout: output cut at %d bytes — this query has no literal anchor, so "+
				"scout cannot rank it.\nNarrow it, or run rg directly to see everything.\n",
			written)
	}
	code := 0
	if err := cmd.Wait(); err != nil {
		if ee, ok := err.(*exec.ExitError); ok {
			code = ee.ExitCode()
		} else {
			code = 2
		}
	}
	if truncated {
		return 0, nil
	}
	return code, nil
}

File: src/executor.rs
//! Runs `rg --json` and folds the stream into per-file records.
//! Memory is O(files), never O(matches): counters stay exact while only a
//! handful of representative lines per file are retained.

use std::io::{BufRead, BufReader};
use std::process::{Command, Stdio};

use crate::ranker;

#[derive(Debug, Clone)]
pub struct Line {
    pub n: u64,
    pub text: String,
    pub score: f32,
    pub is_def: bool,
}

#[derive(Debug)]
pub struct FileHit {
    pub path: String,
    pub matches: u64,
    /// Bit i set when token i of the query was seen anywhere in this file.
    /// Tracked over every match line, so coverage is exact even though only a
    /// few lines are retained.
    pub covered: u32,
    /// Best representative lines, highest score first.
    pub best: Vec<Line>,
    /// Best definition-shaped line seen in this file, if any. Tracked over the
    /// whole stream so a definition can never be lost to the retention cap.
    pub def: Option<Line>,
    pub score: f32,
}

#[derive(Debug, Default)]
pub struct Outcome {
    pub files: Vec<FileHit>,
    pub total_matches: u64,
    pub total_files: u64,
    /// Byte size of the equivalent `rg -n --no-heading` output. This is the
    /// faithful baseline the `never_worse` invariant compares against.
    pub faithful_bytes: u64,
    pub rg_exit: i32,
}

const KEEP_PER_FILE: usize = 3;
const MAX_LINE_CHARS: usize = 600;

fn strip_prefix(p: &str) -> &str {
    p.strip_prefix("./").unwrap_or(p)
}

pub fn run(rg_args: &[String], tokens: &[String], detect_defs: bool) -> std::io::Result<Outcome> {
    let mut cmd = Command::new("rg");
    cmd.arg("--json");
    cmd.args(rg_args);
    cmd.stdout(Stdio::piped());
    cmd.stderr(Stdio::piped());

    let mut child = cmd.spawn()?;
    let stdout = child.stdout.take().expect("piped");
    let reader = BufReader::new(stdout);

    let mut out = Outcome::default();
    let mut cur: Option<FileHit> = None;

    for line in reader.lines() {
        let line = line?;
        let v: serde_json::Value = match serde_json::from_str(&line) {
            Ok(v) => v,
            Err(_) => continue,
        };
        let ty = v.get("type").and_then(|t| t.as_str()).unwrap_or("");
        let data = match v.get("data") {
            Some(d) => d,
            None => continue,
        };

        match ty {
            "begin" => {
                let path = data
                    .get("path")
                    .and_then(|p| p.get("text"))
                    .and_then(|t| t.as_str())
                    .unwrap_or("")
                    .to_string();
                cur = Some(FileHit {
                    path: strip_prefix(&path).to_string(),
                    matches: 0,
                    covered: 0,
                    best: Vec::new(),
                    def: None,
                    score: 0.0,
                });
            }
            "match" => {
                let f = match cur.as_mut() {
                    Some(f) => f,
                    None => continue,
                };
                // Non-UTF8 lines arrive as `bytes` (base64); we count them but
                // never try to display them.
                let text = match data.get("lines").and_then(|l| l.get("text")).and_then(|t| t.as_str()) {
                    Some(t) => t,
                    None => {
                        f.matches += 1;
                        out.total_matches += 1;
                        continue;
                    }
                };
                let n = data.get("line_number").and_then(|x| x.as_u64()).unwrap_or(0);

                f.matches += 1;
                out.total_matches += 1;
                out.faithful_bytes += (f.path.len() + 1 + digits(n) + 1 + text.len()) as u64;

                let trimmed = text.trim_end_matches(['\n', '\r']).trim();
                if trimmed.is_empty() {
                    continue;
                }
                let mut display: String = trimmed.chars().take(MAX_LINE_CHARS).collect();
                if display.chars().count() < trimmed.chars().count() {
                    display.push('…');
                }

                let lower = display.to_lowercase();
                for (i, t) in tokens.iter().take(32).enumerate() {
                    if lower.contains(t.as_str()) {
                        f.covered |= 1 << i;
                    }
                }

                let (score, mut is_def) = ranker::line_score(&display, tokens);
                is_def &= detect_defs;
                let l = Line { n, text: display, score, is_def };

                if is_def && f.def.as_ref().map_or(true, |d| l.score > d.score) {
                    f.def = Some(l.clone());
                }
                if f.best.len() < KEEP_PER_FILE {
                    f.best.push(l);
                    f.best.sort_by(|a, b| b.score.total_cmp(&a.score));
                } else if let Some(worst) = f.best.last() {
                    if l.score > worst.score {
                        f.best.pop();
                        f.best.push(l);
                        f.best.sort_by(|a, b| b.score.total_cmp(&a.score));
                    }
                }
            }
            "end" => {
                if let Some(f) = cur.take() {
                    if f.matches > 0 {
                        out.total_files += 1;
                        out.files.push(f);
                    }
                }
            }
            _ => {}
        }
    }

    let status = child.wait()?;
    out.rg_exit = status.code().unwrap_or(2);
    if out.rg_exit == 2 {
        if let Some(mut err) = child.stderr.take() {
            let mut buf = String::new();
            use std::io::Read;
            let _ = err.read_to_string(&mut buf);
            if !buf.trim().is_empty() {
                eprint!("{buf}");
            }
        }
    }
    Ok(out)
}

/// Streams `rg` verbatim to stdout, so output is byte-identical to plain rg.
/// Used for passthrough and for the `never_worse` fallback.
///
/// One deviation, and it is deliberate: the stream stops at `cap` bytes. An
/// unanchored regex like `[a-z]{2,}` produces 18 MB on a mid-size repo, and
/// handing that to an agent is worse than any ranking mistake. The cut is
/// stated on stderr with the exact command to get the rest.
pub fn passthrough(rg_args: &[String], cap: usize) -> std::io::Result<i32> {
    use std::io::Write;

    let mut child = Command::new("rg")
        .args(rg_args)
        .stdout(Stdio::piped())
        .spawn()?;
    let mut reader = BufReader::new(child.stdout.take().expect("piped"));
    let stdout = std::io::stdout();
    let mut out = stdout.lock();

    let mut written = 0usize;
    let mut truncated = false;
    let mut buf = String::new();
    loop {
        buf.clear();
        match reader.read_line(&mut buf) {
            Ok(0) => break,
            Ok(_) => {
                if written + buf.len() > cap {
                    truncated = true;
                    break;
                }
                if out.write_all(buf.as_bytes()).is_err() {
                    break;
                }
                written += buf.len();
            }
            Err(_) => break,
        }
    }
    let _ = out.flush();

    if truncated {
        let _ = child.kill();
        eprintln!(
            "\nscout: output cut at {written} bytes — this query has no literal anchor, so \
             scout cannot rank it.\nNarrow it, or run rg directly to see everything."
        );
    }
    let status = child.wait()?;
    Ok(if truncated { 0 } else { status.code().unwrap_or(2) })
}

fn digits(mut n: u64) -> usize {
    if n == 0 {
        return 1;
    }
    let mut d = 0;
    while n > 0 {
        d += 1;
        n /= 10;
    }
    d
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn digit_count() {
        assert_eq!(digits(0), 1);
        assert_eq!(digits(9), 1);
        assert_eq!(digits(10), 2);
        assert_eq!(digits(1234), 4);
    }

    #[test]
    fn dot_slash_stripped() {
        assert_eq!(strip_prefix("./src/a.ts"), "src/a.ts");
        assert_eq!(strip_prefix("src/a.ts"), "src/a.ts");
    }
}
